File: sdk-go/examples/quickstart.go
// Runnable end-to-end example against a real hosted control-plane.
//
//	BOXXKITE_BASE_URL=https://your-control-plane BOXXKITE_API_KEY=bxk_live_... \
//	    go run ./examples/quickstart.go
package main

import (
	"context"
	"errors"
	"fmt"
	"log"
	"os"
	"strings"

	boxxkite "github.com/EvAlssment/boxxkite/sdk-go"
)

func main() {
	baseURL := os.Getenv("BOXXKITE_BASE_URL")
	apiKey := os.Getenv("BOXXKITE_API_KEY")
	if baseURL == "" || apiKey == "" {
		fmt.Fprintln(os.Stderr, "Set BOXXKITE_BASE_URL and BOXXKITE_API_KEY first.")
		os.Exit(1)
	}

	client, err := boxxkite.NewClient(baseURL, apiKey)
	if err != nil {
		log.Fatal(err)
	}

	ctx := context.Background()

	account, err := client.Account(ctx)
	if err != nil {
		exitOnError(err)
	}
	fmt.Printf("Signed in as %s\n", account.Email)

	usage, err := client.Usage(ctx)
	if err != nil {
		exitOnError(err)
	}
	fmt.Printf(
		"Usage: %.1f/%.1f sandbox-hours, %d/%d concurrent\n",
		usage.MonthlySandboxHoursUsed, usage.MonthlySandboxHoursLimit,
		usage.ConcurrentSandboxes, usage.ConcurrentSandboxesLimit,
	)

	req := boxxkite.CreateSandboxRequest{Label: boxxkite.Ptr("sdk-go-quickstart")}
	err = client.WithSandbox(ctx, req, func(sb *boxxkite.Session) error {
		fmt.Printf("Created sandbox %s\n", sb.ID)

		result, err := sb.Exec(ctx, "python3 -c 'print(1 + 1)'", nil)
		if err != nil {
			return err
		}
		fmt.Printf("exec result: %s\n", strings.TrimSpace(result.Stdout))

		if _, err := sb.FileCreate(ctx, "hello.txt", "hello from boxxkite-client (go)\n", nil); err != nil {
			return err
		}
		viewed, err := sb.View(ctx, "hello.txt", nil)
		if err != nil {
			return err
		}
		fmt.Printf("file contents: %s\n", strings.TrimSpace(viewed.Content))
		return nil
	})
	if err != nil {
		exitOnError(err)
	}
	fmt.Println("Sandbox destroyed.")
}

func exitOnError(err error) {
	var apiErr *boxxkite.APIError
	if errors.As(err, &apiErr) {
		fmt.Fprintf(os.Stderr, "API error: %s [%s]\n", apiErr.Message, apiErr.Code)
		os.Exit(1)
	}
	log.Fatal(err)
}

File: sdk-rust/examples/quickstart.rs
//! Runnable end-to-end example against a real hosted control-plane.
//!
//! ```bash
//! BOXKITE_BASE_URL=https://your-control-plane BOXKITE_API_KEY=bxk_live_... \
//!     cargo run --example quickstart
//! ```

use std::process::ExitCode;

use boxkite_client::{BoxkiteError, Client, CreateSandboxOptions, ExecOptions};

#[tokio::main]
async fn main() -> ExitCode {
    let Ok(base_url) = std::env::var("BOXKITE_BASE_URL") else {
        eprintln!("Set BOXKITE_BASE_URL and BOXKITE_API_KEY first.");
        return ExitCode::FAILURE;
    };
    let Ok(api_key) = std::env::var("BOXKITE_API_KEY") else {
        eprintln!("Set BOXKITE_BASE_URL and BOXKITE_API_KEY first.");
        return ExitCode::FAILURE;
    };

    match run(base_url, api_key).await {
        Ok(()) => ExitCode::SUCCESS,
        Err(BoxkiteError::Api { code, message, .. }) => {
            eprintln!("API error: {message} [{code}]");
            ExitCode::FAILURE
        }
        Err(err) => {
            eprintln!("error: {err}");
            ExitCode::FAILURE
        }
    }
}

async fn run(base_url: String, api_key: String) -> Result<(), BoxkiteError> {
    let client = Client::new(base_url, api_key)?;

    let sandbox = client
        .create_sandbox(CreateSandboxOptions::new().label("sdk-quickstart"))
        .await?;
    println!("Created sandbox {}", sandbox.id);

    // Best-effort teardown even if a later step fails.
    let result = run_workload(&client, &sandbox.id).await;
    client.destroy_sandbox(&sandbox.id).await?;
    println!("Sandbox destroyed.");

    result
}

async fn run_workload(client: &Client, session_id: &str) -> Result<(), BoxkiteError> {
    let exec_result = client
        .exec(session_id, "python3 -c 'print(1 + 1)'", ExecOptions::new())
        .await?;
    println!("exec result: {}", exec_result.stdout.trim());

    client
        .file_create(
            session_id,
            "hello.txt",
            "hello from boxkite-client\n",
            Default::default(),
        )
        .await?;
    let viewed = client
        .view(session_id, "hello.txt", Default::default())
        .await?;
    println!("file contents: {}", viewed.content.trim());

    Ok(())
}
